### PR TITLE
feature(rework-bugs): Added functionality to update quality investiga…

### PR DESCRIPTION
**Pull request by ds-mwesener**
…tions.

### CODE REVIEW

File: backend/src/integration/groovy/org/eclipse/tractusx/traceability/infrastructure/edc/blackbox/EdcControllerIT.groovy [hunk 1]
```diff
@@ -1,303 +1,136 @@
...
-		}
-
-
+		assertNotificationsSize(1)
+		assertInvestigationsSize(1)
+		assertInvestigationStatus(InvestigationStatus.RECEIVED)
 	}
 
-	def "should call the /qualitynotifications/receive api with wrong requestobject "() {
+	def "should not create a investigation on API callback /qualitynotifications/receive bad request wrong bpn"() {
```
Comment (mkanal) [inline backend/src/integration/groovy/org/eclipse/tractusx/traceability/infrastructure/edc/blackbox/EdcControllerIT.groovy:49]: please explain more in details what "
bad request wrong bpn" is

File: backend/src/integration/groovy/org/eclipse/tractusx/traceability/infrastructure/edc/blackbox/EdcControllerIT.groovy [hunk 2]
```diff
@@ -1,303 +1,136 @@
...
-			[], "BPNL00000003AXS3", InvestigationStatus.RECEIVED, InvestigationSide.RECEIVER, "", "some-description", Instant.now())
-
-		InvestigationEntity investigationEntity = storedInvestigationFullObject(investigation)
-
-		NotificationEntity notificationEntity = storedNotification(notification)
-
-		notificationEntity.setInvestigation(investigationEntity)
-		storedNotification(notificationEntity)
-
+		String notificationJson = readFile("edc_notification_wrong_bpn.json")
```
Comment (mkanal) [inline backend/src/integration/groovy/org/eclipse/tractusx/traceability/infrastructure/edc/blackbox/EdcControllerIT.groovy:51]:
```suggestion
		final String notificationJson = readFile("edc_notification_wrong_bpn.json")
```

File: backend/src/main/java/org/eclipse/tractusx/traceability/infrastructure/edc/blackbox/model/EDCNotificationFactory.java [hunk 3]
```diff
@@ -22,40 +22,46 @@
...
+        }
+        EDCNotificationHeader header = new EDCNotificationHeader(
+                notification.getId(),
+                notification.getSenderBpnNumber(),
+                senderEDC,
+                notification.getReceiverBpnNumber(),
+                NotificationType.QMINVESTIGATION.getValue(),
+                notification.getSeverity() != null ? notification.getSeverity().name() : Severity.MINOR.name(),
+                notification.getNotificationReferenceId(),
+                notification.getInvestigationStatus().name(),
```
Comment (mkanal) [inline backend/src/main/java/org/eclipse/tractusx/traceability/infrastructure/edc/blackbox/model/EDCNotificationFactory.java:50]: what will happen if status is null? or is this a enum?

File: backend/src/main/java/org/eclipse/tractusx/traceability/infrastructure/edc/blackbox/model/EDCNotificationFactory.java [hunk 4]
```diff
@@ -22,40 +22,46 @@
...
+    }
+
+    public static EDCNotification createQualityInvestigation(String senderEDC, Notification notification) {
+        String targetDate = "";
+        if (notification.getTargetDate() == null) {
+            targetDate = Instant.now().toString();
+        } else {
+            targetDate = notification.getTargetDate().toString();
+        }
+        EDCNotificationHeader header = new EDCNotificationHeader(
```
Comment (mkanal) [inline backend/src/main/java/org/eclipse/tractusx/traceability/infrastructure/edc/blackbox/model/EDCNotificationFactory.java:42]: Strongly recomment to use the BuilderPattern here. Target date should be inside the EDCNotificationHeader if null is passed to have a common behavior for EDCNotificationHeader

File: backend/src/main/java/org/eclipse/tractusx/traceability/infrastructure/edc/blackbox/model/EDCNotificationFactory.java [hunk 5]
```diff
@@ -22,40 +22,46 @@
...
-		return notification.getAffectedParts().stream()
-			.map(AffectedPart::assetId)
-			.collect(Collectors.toList());
-	}
+    private EDCNotificationFactory() {
+    }
+
+    public static EDCNotification createQualityInvestigation(String senderEDC, Notification notification) {
+        String targetDate = "";
+        if (notification.getTargetDate() == null) {
```
Comment (mkanal) [inline backend/src/main/java/org/eclipse/tractusx/traceability/infrastructure/edc/blackbox/model/EDCNotificationFactory.java:37]: Target date is a String so StringUtils.isEmpty() would be an apporach

File: backend/src/main/java/org/eclipse/tractusx/traceability/infrastructure/jpa/investigation/JpaInvestigationRepository.java [hunk 6]
```diff
@@ -38,5 +39,6 @@ public interface JpaInvestigationRepository extends JpaRepository<InvestigationE
 	long countAllByStatusEquals(InvestigationStatus status);
 	long countAllBySideEquals(InvestigationSide investigationSide);
 	long countAllByStatusIn(Set<InvestigationStatus> status);
-	Optional<InvestigationEntity> findByNotificationsNotificationReferenceId(String notificationId);
+	@Query("SELECT i FROM InvestigationEntity i JOIN i.notifications n WHERE n.id = :notificationId")
```
Comment (mkanal) [inline backend/src/main/java/org/eclipse/tractusx/traceability/infrastructure/jpa/investigation/JpaInvestigationRepository.java:42]: recommend to use comprehensible names instead of i and n. This is not clear and hard to understand.

File: backend/src/main/java/org/eclipse/tractusx/traceability/infrastructure/jpa/notification/NotificationEntity.java [hunk 7]
```diff
@@ -72,13 +72,14 @@ public NotificationEntity() {
 
 	public NotificationEntity(InvestigationEntity investigation, String senderBpnNumber,
 							  String receiverBpnNumber, List<AssetEntity> assets, String notificationReferenceId,
-							  Instant targetDate) {
+							  Instant targetDate, Severity severity) {
```
Comment (mkanal) [inline backend/src/main/java/org/eclipse/tractusx/traceability/infrastructure/jpa/notification/NotificationEntity.java:75]: final

Comment (mkanal) [reply backend/src/main/java/org/eclipse/tractusx/traceability/infrastructure/jpa/notification/NotificationEntity.java:75]: Stringy recommend to use the BuilderPattern here as

File: backend/src/main/java/org/eclipse/tractusx/traceability/investigations/adapters/jpa/PersistentInvestigationsRepository.java [hunk 8]
```diff
@@ -29,12 +29,7 @@
 import org.eclipse.tractusx.traceability.infrastructure.jpa.investigation.JpaInvestigationRepository;
 import org.eclipse.tractusx.traceability.infrastructure.jpa.notification.JpaNotificationRepository;
 import org.eclipse.tractusx.traceability.infrastructure.jpa.notification.NotificationEntity;
-import org.eclipse.tractusx.traceability.investigations.domain.model.AffectedPart;
-import org.eclipse.tractusx.traceability.investigations.domain.model.Investigation;
-import org.eclipse.tractusx.traceability.investigations.domain.model.InvestigationId;
-import org.eclipse.tractusx.traceability.investigations.domain.model.InvestigationSide;
-import org.eclipse.tractusx.traceability.investigations.domain.model.InvestigationStatus;
-import org.eclipse.tractusx.traceability.investigations.domain.model.Notification;
+import org.eclipse.tractusx.traceability.investigations.domain.model.*;
```
Comment (mkanal) [inline backend/src/main/java/org/eclipse/tractusx/traceability/investigations/adapters/jpa/PersistentInvestigationsRepository.java:32]: Use no wildcard in import.

File: backend/src/main/java/org/eclipse/tractusx/traceability/investigations/adapters/jpa/PersistentInvestigationsRepository.java [hunk 9]
```diff
@@ -143,8 +138,8 @@ public long countPendingInvestigations() {
 	}
 
 	@Override
-	public Optional<Investigation> findByNotificationReferenceId(String notificationId) {
-		return investigationRepository.findByNotificationsNotificationReferenceId(notificationId)
+	public Optional<Investigation> findByNotificationId(String notificationId) {
+		return investigationRepository.findByNotificationsNotificationId(notificationId)
```
Comment (mkanal) [inline backend/src/main/java/org/eclipse/tractusx/traceability/investigations/adapters/jpa/PersistentInvestigationsRepository.java:142]: this method is not clear findByNotificationId?

File: backend/src/main/java/org/eclipse/tractusx/traceability/investigations/adapters/jpa/PersistentInvestigationsRepository.java [hunk 10]
```diff
@@ -165,16 +160,36 @@ private void update(InvestigationEntity investigationEntity, Investigation inves
...
-		investigationEntity.getNotifications().forEach(notification -> {
-			investigation.getNotification(notification.getId()).ifPresent(data -> {
-				update(notification, data);
-			});
-		});
+		// Persist existing notifications
+		investigationEntity.getNotifications()
+			.forEach(notification -> investigation.getNotification(notification.getId())
+				.ifPresent(data -> update(notification, data)));
+		List<Notification> notifications = investigation.getNotifications();
```
Comment (mkanal) [inline backend/src/main/java/org/eclipse/tractusx/traceability/investigations/adapters/jpa/PersistentInvestigationsRepository.java:167]: final

File: backend/src/main/java/org/eclipse/tractusx/traceability/investigations/adapters/jpa/PersistentInvestigationsRepository.java [hunk 11]
```diff
@@ -165,16 +160,36 @@ private void update(InvestigationEntity investigationEntity, Investigation inves
...
+		List<NotificationEntity> notificationEntities = investigationEntity.getNotifications();
+
+		// Persist new notifications
+		List<Notification> notPersistedNotifications =
```
Comment (mkanal) [inline backend/src/main/java/org/eclipse/tractusx/traceability/investigations/adapters/jpa/PersistentInvestigationsRepository.java:171]: Too much business code for one method - extract to own methods.

File: backend/src/main/java/org/eclipse/tractusx/traceability/investigations/domain/model/Investigation.java [hunk 12]
```diff
@@ -129,24 +129,24 @@ public String getDescription() {
...
+                Collections.unmodifiableList(assetIds),
+                investigationSide,
+                new InvestigationReason(
+                        closeReason,
+                        acceptReason,
+                        declineReason
+                ),
+                sendTo,
+                notifications.entrySet().stream().findFirst().map(Map.Entry::getValue).map(Notification::getSeverity).orElse(Severity.MINOR).name(),
+                notifications.entrySet().stream().findFirst().map(Map.Entry::getValue).map(Notification::getTargetDate).map(Instant::toString).orElse(Instant.now().toString()));
```
Comment (mkanal) [inline backend/src/main/java/org/eclipse/tractusx/traceability/investigations/domain/model/Investigation.java:148]: Add comment what is done here // this is required.

File: backend/src/main/java/org/eclipse/tractusx/traceability/investigations/domain/model/Investigation.java [hunk 13]
```diff
@@ -129,24 +129,24 @@ public String getDescription() {
...
-				closeReason,
-				acceptReason,
-				declineReason
-			),
-			sendTo,
-			notifications.entrySet().stream().findFirst().map(Map.Entry::getValue).map(Notification::getSeverity).orElse(Severity.MINOR).name()
-		);
-	}
+    public InvestigationData toData() {
+        return new InvestigationData(
```
Comment (mkanal) [inline backend/src/main/java/org/eclipse/tractusx/traceability/investigations/domain/model/Investigation.java:133]: Use Builder Pattern here

File: backend/src/main/java/org/eclipse/tractusx/traceability/investigations/domain/model/Investigation.java [hunk 14]
```diff
@@ -208,6 +225,16 @@ private void changeStatusTo(InvestigationStatus to) {
 		this.investigationStatus = to;
 	}
 
+	private void changeStatusToWithoutNotifications(InvestigationStatus to) {
+		boolean transitionAllowed = investigationStatus.transitionAllowed(to);
+
+		if (!transitionAllowed) {
```
Comment (mkanal) [inline backend/src/main/java/org/eclipse/tractusx/traceability/investigations/domain/model/Investigation.java:231]: ownmethod isTransitionAllowed(to)

File: backend/src/test/java/org/eclipse/tractusx/traceability/investigations/domain/service/InvestigationsReceiverServiceTest.java [hunk 15]
```diff
@@ -123,143 +81,5 @@ void testHandleNotificationReceiverCallbackValidSentNotification() {
 		Mockito.verify(mockRepository).save(investigationTestData);
 	}
```
Comment (mkanal) [inline backend/src/test/java/org/eclipse/tractusx/traceability/investigations/domain/service/InvestigationsReceiverServiceTest.java:83]: Why was this removed?

File: backend/src/test/java/org/eclipse/tractusx/traceability/investigations/domain/service/InvestigationsPublisherServiceTest.java [hunk 16]
```diff
@@ -114,7 +118,145 @@ void testSendInvestigationSuccessful() {
...
+	void testUpdateInvestigationInvalidBpnMismatch() {
+
+		// Given
+		BPN bpn = BPN.of("applicationBPN");
+		Long investigationIdRaw = 1L;
+		InvestigationStatus status = InvestigationStatus.ACKNOWLEDGED;
+		String reason = "the update reason";
+
+		List<AffectedPart> affectedParts = List.of(new AffectedPart("partId"));
+		Notification notification = new Notification(
```
Comment (mkanal) [inline backend/src/test/java/org/eclipse/tractusx/traceability/investigations/domain/service/InvestigationsPublisherServiceTest.java:231]: Recommendation to use Builder patterns for such complex objects

File: backend/src/main/java/org/eclipse/tractusx/traceability/investigations/domain/service/NotificationsService.java [hunk 17]
```diff
@@ -45,16 +50,30 @@ public NotificationsService(InvestigationsEDCFacade edcFacade, InvestigationsRep
 	}
 
 	@Async(value = AssetsAsyncConfig.UPDATE_NOTIFICATION_EXECUTOR)
-	public void updateAsync(Notification notification) {
+	public void updateAsync(Notification notification, boolean isReceiver) {
```
Comment (mkanal) [inline backend/src/main/java/org/eclipse/tractusx/traceability/investigations/domain/service/NotificationsService.java:53]: Why is receiver and sender property not bound to Notification object? Why is the notification not extended with the direction?

File: backend/src/main/java/org/eclipse/tractusx/traceability/investigations/domain/service/NotificationsService.java [hunk 18]
```diff
@@ -37,6 +40,8 @@ public class NotificationsService {
 	private final InvestigationsEDCFacade edcFacade;
 	private final InvestigationsRepository repository;
 	private final EDCUrlProvider edcUrlProvider;
+	private static final Logger logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
```
Comment (mkanal) [inline backend/src/main/java/org/eclipse/tractusx/traceability/investigations/domain/service/NotificationsService.java:43]: use static imports. Do this similar in all classes that developer could rely on a common approach: private static final Logger logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());

Comment (mkanal) [inline backend/src/main/java/org/eclipse/tractusx/traceability/investigations/domain/service/NotificationsService.java:43]: lookup == null?

File: backend/src/main/java/org/eclipse/tractusx/traceability/investigations/domain/service/InvestigationsReceiverService.java [hunk 19]
```diff
@@ -32,105 +32,99 @@
...
+        repository.save(investigation);
+        logger.info("Stored received notification in investigation {}", investigation);
+    }
+
+    private void receiveUpdateInvestigation(EDCNotification edcNotification, InvestigationStatus investigationStatus) {
+        logger.info("receiveUpdateInvestigation with status {}", investigationStatus);
+        Notification notification = notificationMapper.toReceiverNotification(edcNotification, investigationStatus);
+        Investigation investigation = investigationsReadService.loadInvestigationByNotificationId(edcNotification.getRelatedNotificationId());
+
+        switch (investigationStatus) {
```
Comment (mkanal) [inline backend/src/main/java/org/eclipse/tractusx/traceability/investigations/domain/service/InvestigationsReceiverService.java:103]: This looks quite strange.

File: backend/src/main/java/org/eclipse/tractusx/traceability/investigations/domain/service/InvestigationsReceiverService.java [hunk 20]
```diff
@@ -32,105 +32,99 @@
...
+    public void handleNotificationReceiverCallback(EDCNotification edcNotification) {
+        logger.info("Received notification response with id {}", edcNotification.getNotificationId());
+
+        BPN recipientBPN = BPN.of(edcNotification.getRecipientBPN());
+
+        validateNotificationReceiverCallback(edcNotification);
+
+        InvestigationStatus investigationStatus = edcNotification.convertInvestigationStatus();
+
+        switch (investigationStatus) {
```
Comment (mkanal) [inline backend/src/main/java/org/eclipse/tractusx/traceability/investigations/domain/service/InvestigationsReceiverService.java:72]: again the switch that looks very strange. Could that be extracted?

File: backend/src/main/java/org/eclipse/tractusx/traceability/investigations/domain/service/InvestigationsReceiverService.java [hunk 21]
```diff
@@ -32,105 +32,99 @@
...
+        this.investigationsReadService = investigationsReadService;
+        this.notificationMapper = notificationMapper;
+        this.investigationMapper = investigationMapper;
+        this.traceabilityProperties = traceabilityProperties;
+    }
+
+    public void handleNotificationReceiverCallback(EDCNotification edcNotification) {
+        logger.info("Received notification response with id {}", edcNotification.getNotificationId());
+
+        BPN recipientBPN = BPN.of(edcNotification.getRecipientBPN());
```
Comment (mkanal) [inline backend/src/main/java/org/eclipse/tractusx/traceability/investigations/domain/service/InvestigationsReceiverService.java:66]: why is recipientBPN not checked if null? What will haben if it is null? This should be handeld or even logged.

File: backend/src/main/java/org/eclipse/tractusx/traceability/investigations/domain/service/InvestigationsReadService.java [hunk 22]
```diff
@@ -76,8 +76,8 @@ public Investigation loadInvestigation(InvestigationId investigationId) {
 			.orElseThrow(() -> new InvestigationNotFoundException(investigationId));
 	}
 
-	public Investigation loadInvestigationByNotificationReferenceId(String notificationId) {
-		return repository.findByNotificationReferenceId(notificationId)
+	public Investigation loadInvestigationByNotificationId(String notificationId) {
```
Comment (mkanal) [inline backend/src/main/java/org/eclipse/tractusx/traceability/investigations/domain/service/InvestigationsReadService.java:79]: MethodNames differ: 
loadInvestigationByNotificationId
findByNotificationId
This should having the same naming for all service methods.

load/find(Entity)By{Criteria}

File: backend/src/main/java/org/eclipse/tractusx/traceability/investigations/domain/service/InvestigationsReceiverService.java (hunk 19 above)
Comment (ds-mwesener) [reply backend/src/main/java/org/eclipse/tractusx/traceability/investigations/domain/service/InvestigationsReceiverService.java:103]: Why? Do you mean that we load the notification based on the relatedNotificationId? This is to have a relation between the notifications which will be exchanged between two instances of the application running.

**Review by mkanal — APPROVED** [review]
